flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/v2/IcebergFilesCommitter.java

```diff
@@ -0,0 +1,261 @@
+  private static final String FLINK_JOB_ID = "flink.job-id";
+
+  // The max checkpoint id we've committed to iceberg table. As the flink's checkpoint is always increasing, so we could
+  // correctly commit all the data files whose checkpoint id is greater than the max committed one to iceberg table, for
+  // avoiding committing the same data files twice. This id will be attached to iceberg's meta when committing the
+  // iceberg transaction.
+  private static final String MAX_COMMITTED_CHECKPOINT_ID = "flink.max-committed-checkpoint-id";
+  static final String MAX_CONTINUOUS_EMPTY_COMMITS = "flink.max-continuous-empty-commits";
+
+  // TableLoader to load iceberg table lazily.
+  private final TableLoader tableLoader;
+  private final boolean replacePartitions;
+  private final Map<String, String> snapshotProperties;
+
+  // It will have an unique identifier for one job.
+  private final transient Table table;
+  private transient long maxCommittedCheckpointId;
+  private transient int continuousEmptyCheckpoints;
+  private final transient int maxContinuousEmptyCommits;
+
+
+  private final transient ExecutorService workerPool;
+
+  public IcebergFilesCommitter(TableLoader tableLoader,
+                               boolean replacePartitions,
+                               Map<String, String> snapshotProperties,
+                               Integer workerPoolSize) {
+
+    this.tableLoader = tableLoader;
+    this.replacePartitions = replacePartitions;
+    this.snapshotProperties = snapshotProperties;
+
+    this.workerPool =
+        ThreadPools.newWorkerPool("iceberg-worker-pool-" + Thread.currentThread().getName(), workerPoolSize);
+
+    // Open the table loader and load the table.
+    this.tableLoader.open();
+    this.table = tableLoader.loadTable();
+
+    maxContinuousEmptyCommits = PropertyUtil.propertyAsInt(table.properties(), MAX_CONTINUOUS_EMPTY_COMMITS, 10);
+    Preconditions.checkArgument(
+        maxContinuousEmptyCommits > 0,
+        MAX_CONTINUOUS_EMPTY_COMMITS + " must be positive");
+    this.maxCommittedCheckpointId = INITIAL_CHECKPOINT_ID;
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<IcebergFlinkCommittable>> requests)
+      throws IOException, InterruptedException {
+    int dataFilesNumRestored = 0;
+    int deleteFilesNumRestored = 0;
+
+    int dataFilesNumStored = 0;
+    int deleteFilesNumStored = 0;
+
+    long checkpointId = 0;
+    Collection<CommitRequest<IcebergFlinkCommittable>> restored = Lists.newArrayList();
+    Collection<CommitRequest<IcebergFlinkCommittable>> store = Lists.newArrayList();
+
+    for (CommitRequest<IcebergFlinkCommittable> request : requests) {
+      if (request.getNumberOfRetries() > 0) {
+        if (request.getCommittable().checkpointId() > maxCommittedCheckpointId) {
+          checkpointId = request.getCommittable().checkpointId();
+          restored.add(request);
+          WriteResult committable = request.getCommittable().committable();
```

Comment:
We can extract the same code out of if/else block.

Reply:
done